rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/WorkspaceInfo.rsc

```diff
@@ -0,0 +1,404 @@
+}
+@bootstrapParser
+module lang::rascal::lsp::refactor::WorkspaceInfo
+
+import Relation;
+
+import analysis::typepal::TModel;
+
+import lang::rascalcore::check::Checker;
+
+import lang::rascal::\syntax::Rascal;
+
+import util::Maybe;
+import util::Reflective;
+
+import lang::rascal::lsp::refactor::Exception;
+import lang::rascal::lsp::refactor::Util;
+
+import List;
+import Location;
+import Map;
+import Set;
+import String;
+
+data CursorKind
+    = use()
+    | def()
+    | typeParam()
+    | collectionField()
+    | moduleName()
+    ;
+
+data Cursor
+    = cursor(CursorKind kind, loc l, str name)
+    ;
+
+alias MayOverloadFun = bool(set[loc] defs, map[loc, Define] defines);
+alias FileRenamesF = rel[loc old, loc new](str newName);
+alias DefsUsesRenames = tuple[set[loc] defs, set[loc] uses, FileRenamesF renames];
+alias ProjectFiles = rel[loc projectFolder, loc file];
+
+/**
+ * This is a subset of the fields from analysis::typepal::TModel, specifically tailored to refactorings.
+ * WorkspaceInfo comes with a set of functions that allow (incrementally) loading information from multiple TModels, and doing (cached) queries on this data.
+ */
+data WorkspaceInfo (
+    // Instance fields
+    // Read-only
+    rel[loc use, loc def] useDef = {},
+    set[Define] defines = {},
+    set[loc] sourceFiles = {},
+    map[loc, Define] definitions = (),
+    map[loc, AType] facts = (),
+    Scopes scopes = (),
+    Paths paths = {},
+    set[loc] projects = {}
+) = workspaceInfo(
+    ProjectFiles() preloadFiles,
+    ProjectFiles() allFiles,
+    set[TModel](ProjectFiles) tmodelsForLocs
+);
+
+WorkspaceInfo loadLocs(WorkspaceInfo ws, ProjectFiles projectFiles) {
+    for (tm <- ws.tmodelsForLocs(projectFiles)) {
+        ws = loadTModel(ws, tm);
+    }
+
+    // In addition to data from the TModel, we keep track of which projects/modules we loaded.
+    ws.sourceFiles += projectFiles.file;
+    ws.projects += projectFiles.projectFolder;
+
+    return ws;
+}
+
+WorkspaceInfo preLoad(WorkspaceInfo ws) {
+    return loadLocs(ws, ws.preloadFiles());
+}
+
+WorkspaceInfo loadWorkspace(WorkspaceInfo ws) {
+    return loadLocs(ws, ws.allFiles());
+}
+
+WorkspaceInfo loadTModel(WorkspaceInfo ws, TModel tm) {
+    try {
+        throwAnyErrors(tm);
+    } catch set[Message] errors: {
+        throw unsupportedRename("Cannot rename: some files in workspace have errors.\n<toString(errors)>", issues={<(error.at ? |unknown:///|), error.msg> | error <- errors});
+    }
+
+    ws.useDef      += tm.useDef;
+    ws.defines     += tm.defines;
+    ws.definitions += tm.definitions;
+    ws.facts       += tm.facts;
+    ws.scopes      += tm.scopes;
+    ws.paths       += tm.paths;
+
+    return ws;
+}
+
+loc getProjectFolder(WorkspaceInfo ws, loc l) {
+    if (project <- ws.projects, isPrefixOf(project, l)) {
+        return project;
+    }
+
+    throw "Could not find project containing <l>";
+}
+
+@memo{maximumSize(1), expireAfter(minutes=(5))}
```

Comment:
After renaming, the WorkspaceInfo will be recalculated (since renaming changes/outdates TPLs). Hence size 1. As long as no renaming succeeds and no modules are changed otherwise, the cache remains valid.